### PR TITLE
Draft: Simultaneous execution of trajectories

### PR DESCRIPTION
**Pull request by cambel**
@felixvd and I have been working on this PR

### Description

This draft PR proposes a method to allow the execution of multiple controllers in several independent trajectory plans at the same time.
So far, MoveIt only supports the execution of one trajectory at a time. If a second trajectory is sent for execution, the first trajectory is aborted and the second one is attempted instead. 

There are two main reasons for this limitation. First, all the action servers are of the type SimpleActionServer which can only serve one goal at a time and immediately reject any old goal as a new one is received. Second, the TrajectoryExecutionManager only supports the execution of one trajectory at a time by default, aborting any previous trajectory as a new one is pushed. 

The first issue is addressed by converting the SimpleActionServers into (General) ActionServers. This allows the server to handle multiple goals at the same time.
The second issue is addressed by implementing a **scheduling procedure** to execute trajectories simultaneously as long as the following conditions are met:

- The controllers needed for executing the trajectory are not being used.
- The trajectory is not in collision with any active trajectory.
- The trajectory is not in collision with the current state. This is assumed in the current trajectory execution scheme as it is expected that the planning covers this validation. For most cases, it works fine, but for simultaneous execution of trajectories, that assumption does not hold anymore, so an extra validation is needed.
- The start state of the trajectory matched that of the current state of the robot.

If the trajectory does not meet the conditions mentioned above, the trajectory is stored in a backlog and periodically checked. The trajectories in the backlog and newly received trajectories are evaluated and executed in a strict FIFO order.

Our use case: Simultaneous control of robots with different sets of instructions that do not have/need to be synchronized in any particular way. See this simple equip/unequip routine where both robots can follow their independent set of instructions without waiting for the other one.

https://user-images.githubusercontent.com/3798796/129310195-fa0787bf-6942-4447-a33c-ecbbbe054efd.mp4

There are some issues pending to be fixed:

- Re-enable feedback for the action servers.
- Simultaneous planning seems to crash MoveIt/move_group. It is a non-deterministic bug that we are still trying to understand. I am trying to debug [here](https://github.com/cambel/moveit/tree/tem-debug). The problem is not uncommon, it happens during the planning, as we do plan_only and then execute the plans. It even happens with only one robot planning at a time. The crash happens when sending a solution plan to the goal_handle (of the ActionServer), for whatever reason the executed trajectory of the action response (`MoveGroupResult`) is corrupted because something cannot be accessed `Cannot access memory at address`.
- Unless I am mistaken, the `execute_trajectory_service` also needs to be changed over to simultaneous execution
- There are no tests for the new behavior yet

Comments and contributions are highly appreciated!

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by felixvd** [comment]
Fixes https://github.com/ros-planning/moveit/issues/2287 

@JafarAbdi  @henningkayser @tylerjw This is the project I mentioned in the call yesterday and the maintainer meeting. Some issues remain, but we have been running this for a few weeks and it should be ready for testing and feedback from a fresh set of eyes. There are infrequent crashes, probably related to thread safety in `move_action.cpp`. It would be great if someone could have a look.

Some more notes:

- We renamed the functions and variables in TrajectoryExecutionManager to specify their behavior (`blocking` vs `simultaneous`). 
- The class originally had two quasi-redundant behaviors. The logic of the blocking queue could *probably* be removed now, but there are some calls in `plan_execution.cpp` and `plan_with_sensing.cpp` and I am not too familiar with those files. I'd like a second opinion. @v4hn ?
- We added the `group_name` to RobotTrajectory because we need to know it for tracking the active joints in the TEM class.

Since the video above does not do this PR justice (slow PlanningScene update rate and ugly CollisionObjects 💩), I want to add some more to drive home the point how sweet and deserving of enthusiastic reviews this is:

1) Two arms rearranging an objects in their own part of the workspace independently, each planning in their own thread in peace, simultaneously, without a care in the world

https://user-images.githubusercontent.com/4535737/129320447-0a2250c1-ae73-4ce2-9a98-bc07d4c54520.mp4

2) The right arm aligns the screw holes of the bearing housing (using vision) while the left arm picks the tool and a screw from the feeder in the back

https://user-images.githubusercontent.com/4535737/129320216-de784998-a601-4e6d-972f-eab238abcbef.mp4

3) The right arm inserts the shaft while the left arm picks screws and fastens the bearing. Note how close the robots pass each other and how suicidal this would be without the TrajectoryExecutionMonitor ensuring that the trajectories don't collide.

https://user-images.githubusercontent.com/4535737/129320653-82bafb71-7d31-4721-a23f-c114505d46aa.mp4

Anyway, I hope we can get the kinks ironed out soon. I think this PR is in everyone's interest. Thanks!


PS: I originally thought that the collision check between two trajectories could be done more efficiently with [gpu-voxels](https://github.com/fzi-forschungszentrum-informatik/gpu-voxels), so we wasted a bit of time on trying that. That was absolutely premature optimization though, the performance as is seems perfectly fine.

**Review by henningkayser — COMMENTED** [review]
@cambel @felixvd Whoa, this is a lot. Great work so far! I really like the general idea of having an ongoing loop in the TEM that keeps track of any active trajectories and controllers. By validating incoming trajectories with all active instances in the beginning it should be possible to ensure we are at least commanding collision-free trajectories (fingers crossed that the controllers are actually doing what we command ;)). I haven't fully understood all of the new logic yet, I'll follow up with more in-depth and structural feedback. I also still need to make up my mind about how concurrent actions could be handled in general.

So far, my biggest concern is that collision checking between trajectories is inherently unsafe if we do it the way it's implemented right now. Self collisions have reduced padding by default and the resolution of trajectory waypoints is usually not sufficient to really guarantee that different groups don't run into each other.
As already mentioned in one of the comments, I think that we need to:

- Increase padding for self collisions (= collisions between different groups)
- Ensure a minimal joint distance resolution similar to `longest_valid_fragment_section`
- Use a "continuous" collision check that validates the full segment between two consecutive waypoints. I'm not sure if this is currently even supported for self-collisions, but I imagine that this would be the most secure and efficient solution long-term. This could make the minimal joint distance threshold redundant.

Using a very big padding with a reasonable trajectory resolution should be fine for most use cases, but you could still not really rule out collisions for all cases as small joint state jumps can produce larger Cartesian distances.

Do you have a testing setup that you could share for testing?

**Comment by henningkayser** [comment]
Just a thought, it might make sense to split this PR into two parts:

1. Update TrajectoryExecutionManager to support simultaneous trajectories + test directly using MoveItCpp
2. Modify capabilities to support this new feature

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp [hunk 1]
```diff
@@ -231,6 +231,7 @@ void RobotTrajectory::getRobotTrajectoryMsg(moveit_msgs::RobotTrajectory& trajec
 
   std::vector<const moveit::core::JointModel*> onedof;
   std::vector<const moveit::core::JointModel*> mdof;
+  trajectory.group_name = group_ ? group_->getName() : "";
```
[the remark this answers is not in the document]
Comment (cambel) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp:234]: Sure

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h [hunk 2]
```diff
@@ -60,6 +60,8 @@ class BaseFakeController : public moveit_controller_manager::MoveItControllerHan
 protected:
   std::vector<std::string> joints_;
   const ros::Publisher& pub_;
+  moveit_controller_manager::ExecutionStatus status_;
+  std::string name_ = "fake_controllers";
```
[the remark this answers is not in the document]
Comment (cambel) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h:64]: Okay

File: moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.cpp [hunk 3]
```diff
@@ -50,55 +50,36 @@ MoveGroupExecuteTrajectoryAction::MoveGroupExecuteTrajectoryAction() : MoveGroup
...
-  {
-    setExecuteTrajectoryState(MONITOR);
-    context_->trajectory_execution_manager_->execute();
-    moveit_controller_manager::ExecutionStatus status = context_->trajectory_execution_manager_->waitForExecution();
+  
+  // Define a lambda that sets the goal state when the trajectory is completed.
+  // This is called when the trajectory finishes.
+  bool execution_complete = false;
+  auto goalId = goal.getGoalID();
+  // Copy goalId here so that it is still available when this lambda is called later (otherwise it is deleted when the parent function ends)
```
[the remark this answers is not in the document]
Comment (cambel) [reply moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.cpp:79]: No, the function `pushAndExecuteSimultaneous` is not blocking at all, it only pushes a new trajectory to the queue and then returns. The variable `execution_complete` is not actually used at all.

File: moveit_ros/planning/trajectory_execution_manager/include/moveit/trajectory_execution_manager/trajectory_execution_manager.h [hunk 4]
```diff
@@ -130,25 +141,25 @@ class TrajectoryExecutionManager
 
   /// Add a trajectory for future execution. Optionally specify a controller to use for the trajectory. If no controller
   /// is specified, a default is used.
-  bool push(const moveit_msgs::RobotTrajectory& trajectory, const std::string& controller = "");
+  bool pushToBlockingQueue(const moveit_msgs::RobotTrajectory& trajectory, const std::string& controller = "");
```
[the remark this answers is not in the document]
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/include/moveit/trajectory_execution_manager/trajectory_execution_manager.h:144]: I am not sure what you mean by the changes being consistent for all functions. Before this change, `push()` and `pushAndExecute()` were completely different. `push()` is the only one being used for the normal blocking executions of one trajectory at a time. `pushAndExecute()` was apparently not used at all but it was already a prototype of running multiple trajectories continuously, though the logic was broken so it wouldn't work.
Renaming them gives them a better description I guess. Though @felixvd can say more about it.

File: moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.cpp [hunk 5]
```diff
@@ -115,24 +96,58 @@ void MoveGroupExecuteTrajectoryAction::executePath(const moveit_msgs::ExecuteTra
...
       action_res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
     }
-    ROS_INFO_STREAM_NAMED(getName(), "Execution completed: " << status.asString());
-  }
-  else
-  {
-    action_res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
-  }
-}
+    ROS_DEBUG_STREAM_NAMED(name_, "Execution completed: " << status.asString());
```
[the remark this answers is not in the document]
Comment (cambel) [reply moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.cpp:99]: getName() is not define in this scope. I replaced `name_` with LOGNAME as suggested previously.

**Comment by cambel** [comment]
@henningkayser Thanks for the comments. I will slowly check them and update the PR. I created a simple environment to test this PR [here](https://github.com/cambel/moveit_simultaneous_motions_demo) using the Panda dual-arm env.

File: moveit_ros/planning/trajectory_execution_manager/include/moveit/trajectory_execution_manager/trajectory_execution_manager.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixvd) [reply moveit_ros/planning/trajectory_execution_manager/include/moveit/trajectory_execution_manager/trajectory_execution_manager.h:146]: I don't have much to add to that. It seemed like unfinished functionality. The renaming just separates the two approaches.

**Comment by felixvd** [comment]
You can now also test this PR using [the repository we originally wrote it for](https://github.com/o2ac/o2ac-ur).

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonschmeisser) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp:234]: https://github.com/ros-planning/moveit_msgs/pull/133 for reference

**Comment by simonschmeisser** [comment]
I had a quick look at the current state of this PR as well and I think we should factor out at least two additional PRs:

1. changing `name_` to `LOGNAME`
2. add a `PlanningScene::isPathValid` variant that accepts some maximum discretization parameter and subsamples in between trajectory points. If a collision checker supports continuous collision checking more efficiently, it could still use that internally if desired.
3. alternatively we could also provide a separate subsampling algorithm and use the existing `isPathValid`

Then we need to interpolate the current trajectory and check each point with the new `isPathValid`

(repeating myself :wink: , but please check out pre-commit https://moveit.ros.org/documentation/contributing/code/ )

**Comment by cambel** [comment]
Hi @tahsinkose,

1. This PR is for non-unary manipulation systems, however, there is no specific limitation introduced here about how the robots should be represented. However, representing the robots as a single robot model with **different** joint groups was the approach used to test this feature.

> how is this method better than solving the coupled planning problem


2. That is a good question. With a coupled planning system, you can **only** solve planning problems where each robot knows what the others are going to do (how they are going to move). That approach is useful and enough for some applications. However, for more complex applications where you want each robot to solve a different independent task, knowing what the other robots will do beforehand is not possible. So, controlling each joint group independently but simultaneously is necessary.

**Pull request closed without merging**